# Numpy - Domača naloga

Uvozimo modul numpy:

In [2]:
import numpy as np

### 1.Create a null vector of size 10

In [3]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [4]:
matrika = np.zeros(10)
matrika[4] = 1



Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [5]:
np.arange(10, 50, 1)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49])

Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [6]:
Z = np.arange(50)
L = len(Z)
R = np.ones(L, dtype=int)
for i in range(L):
    R[i] = int(Z[L - 1 - i])
R

    



array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,
       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,
       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

In [7]:
#ali
-np.sort(-Z)

array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,
       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,
       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [8]:
np.arange(9).reshape(3, 3)


array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [9]:
a = [1,2,0,0,4,0]

In [10]:
a = np.array(a)
indices = np.arange(len(a))
result = indices[a != 0]
result

array([0, 1, 4])

Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [11]:
id_matr = np.zeros((3, 3))
for i in range(id_matr.shape[0]):
    for j in range(id_matr.shape[0]):
        if j==i:
            id_matr[i, j] = 1
id_matr

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [12]:
np.random.uniform(0, 1, 27).reshape(3, 3, 3)

array([[[0.40667285, 0.95579912, 0.94868545],
        [0.92017362, 0.38687575, 0.29881471],
        [0.93384695, 0.77555788, 0.6016782 ]],

       [[0.58599533, 0.28230826, 0.50479397],
        [0.54586715, 0.66214456, 0.0920065 ],
        [0.15932769, 0.20101108, 0.23702213]],

       [[0.87528876, 0.85915126, 0.48552952],
        [0.11825864, 0.79933326, 0.88910203],
        [0.96444283, 0.64716435, 0.52714387]]])

Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [13]:
ten_by_ten = np.random.uniform(0, 1, 10*10).reshape(10, 10)

In [14]:
Zmin = ten_by_ten.min()
Zmax = ten_by_ten.max()

Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [15]:
border_matrix = np.ones((10, 10))
border_matrix[1:-1, 1:-1] = 0
border_matrix

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

In [16]:
0 * np.nan

nan

In [17]:
#Vse kar delamo z nan-am vrze false (saj je tako misljeno, da se ne bi napake prenasale) https://stackoverflow.com/questions/10034149/why-is-nan-not-equal-to-nan

In [18]:
np.nan == np.nan

False

In [19]:
np.inf > np.nan

False

In [20]:
np.nan - np.nan

nan

In [21]:
0.3 == 3 * 0.1 #floating point error

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [50]:
checkerboard = np.ones((8, 8), dtype=int)
for i in range(0, len(checkerboard)):
    for j in range(0, len(checkerboard[i])):
        if (i+j+2) % 2 == 0:
            checkerboard[i, j] = 0
            

In [51]:
checkerboard

array([[0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0]])

Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [166]:
Z = np.arange(11)

In [167]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [168]:
a = Z[3:8]  #Zakaj ne dela, ce vstavim namesto a = Z[3:8] -> a = Z[(Z>3) & (Z<8)]
a[:] = -a
a

array([-4, -5, -6, -7])

In [169]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [148]:
np.linspace(0, 1, 12)[1:-1]

array([0.09090909, 0.18181818, 0.27272727, 0.36363636, 0.45454545,
       0.54545455, 0.63636364, 0.72727273, 0.81818182, 0.90909091])

Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [171]:
size_ten = np.random.uniform(0, 1, 10)
size_ten.sort()
size_ten

array([0.01316096, 0.01472743, 0.41505642, 0.68875484, 0.74392984,
       0.91113518, 0.91625903, 0.94294189, 0.95745256, 0.9776394 ])

In [ ]:
#Zakaj ne deluje size_ten = np.random.uniform(0, 1, 10).sort()

Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [194]:
A = np.random.randint(0,2,5)

In [195]:
B = np.random.randint(0,2,5)

In [196]:
A == B

array([False,  True, False,  True, False])

In [197]:
np.array_equal(A, B)

False

### 17.Make an array immutable (read-only)

In [17]:
Z = np.zeros(10)

In [199]:
Z.flags.writeable = False

### 18.Create random vector of size 10 and replace the maximum value by 0

In [237]:
R = np.random.random([10])
R[R.max() == R] = 0
R

array([0.36300909, 0.58380126, 0.26029088, 0.34588888, 0.        ,
       0.03160943, 0.25599693, 0.14908586, 0.40334601, 0.68535667])

### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [252]:
Z = np.arange(100)
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [253]:
v = np.random.uniform(0,100)
v

62.91845229922468

In [254]:
Z[(abs(Z - v)).min() == abs((Z - v))]

array([63])

### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [258]:
Z = np.arange(10, dtype=np.int32)

In [259]:
print(Z)

[0 1 2 3 4 5 6 7 8 9]


In [270]:
Z = np.float32(Z)

In [263]:
Z

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float32)

Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [284]:
Z = np.random.randint(0,10,50)
print(Z)
np.bincount(Z)

[1 3 8 3 0 9 1 1 6 3 2 3 6 9 3 0 6 2 6 8 5 2 5 3 5 6 0 5 8 3 3 4 9 0 9 9 3
 8 4 2 3 3 6 2 1 4 4 9 0 8]


array([ 5,  4,  5, 11,  4,  4,  6,  0,  5,  6])

In [285]:
np.where(np.bincount(Z).max() == np.bincount(Z)) #stevilka spodaj je najpogostejsa

(array([3]),)